javascripts/discourse/lib/sidebar-panel.js

```diff
@@ -0,0 +1,145 @@
+import { cached } from "@glimmer/tracking";
+import { computed } from "@ember/object";
+import BaseCustomSidebarSection from "discourse/lib/sidebar/base-custom-sidebar-section";
+import BaseCustomSidebarSectionLink from "discourse/lib/sidebar/base-custom-sidebar-section-link";
+import DiscourseURL from "discourse/lib/url";
+import { getOwnerWithFallback } from "discourse-common/lib/get-owner";
+import { getAbsoluteURL, samePrefix } from "discourse-common/lib/get-url";
+import I18n from "discourse-i18n";
+import { SIDEBAR_DOCS_PANEL } from "../services/docs-sidebar";
+import { normalizeName } from "./utils";
+
+const sidebarPanelClassBuilder = (BaseCustomSidebarPanel) =>
+  class DocsSidebarPanel extends BaseCustomSidebarPanel {
+    key = SIDEBAR_DOCS_PANEL;
+    hidden = true;
+
+    get docsSidebar() {
+      return getOwnerWithFallback(this).lookup("service:docs-sidebar");
+    }
+
+    @cached
+    get sections() {
+      const router = getOwnerWithFallback(this).lookup("service:router");
+
+      return this.docsSidebar.sectionsConfig.map((config) => {
+        return prepareDocsSection({ config, router });
+      });
+    }
+
+    get filterable() {
+      return !this.docsSidebar.loading;
+    }
+  };
+
+export default sidebarPanelClassBuilder;
+
+function prepareDocsSection({ config, router }) {
+  return class extends BaseCustomSidebarSection {
+    #config = config;
+
+    get sectionLinks() {
+      return this.#config.links;
+    }
+
+    get name() {
+      return this.#config.name;
+    }
+
+    get title() {
+      return this.#config.text;
+    }
+
+    get text() {
+      return this.#config.label
+        ? I18n.t(this.#config.label)
+        : this.#config.text;
+    }
+
+    get links() {
+      return this.sectionLinks.map(
+        (sectionLinkData) =>
+          new SidebarDocsSectionLink({
+            data: sectionLinkData,
+            router,
+          })
+      );
+    }
+
+    get displaySection() {
+      return true;
+    }
+
+    get hideSectionHeader() {
+      return this.#config.name === "root";
+    }
+
+    get collapsedByDefault() {
+      return false;
+    }
+  };
+}
+
+class SidebarDocsSectionLink extends BaseCustomSidebarSectionLink {
+  #data;
+
+  constructor({ data, router }) {
+    super(...arguments);
+
+    this.#data = data;
+    this._router = router;
+  }
+
+  @computed("_router.currentRoute", "href")
+  get active() {
+    if (DiscourseURL.isInternal(this.href) && samePrefix(this.href)) {
```

Comment:
I'm surprised you have to do all this since `LinkTo` is supposed to match up the current route and add the `active` class, and you are adding your section links in the normal way for the sidebar :thinking:

Reply:
I wish it worked out of the box.

This little function took me quite a while yesterday to get right.

I'm performing this manually because `LinkTo` only does this automatically when we provide  `route`/`model` arguments. 

In this case, we're providing a `href` argument to `LinkTo` and then, unfortunately, the component won't add the `active` class even if you're in the page.

Furthermore, the topic route also uses two model arguments: `slug` and `id`. Even if the `LinkTo` component was adding the class automatically, it wouldn't work if the slug doesn't match.

As the index is generated manually, sometimes the topic title is updated and the slug in the HTML becomes outdated. 

I'm considering just the ID as it's the only parameter that matters in this case.

When Hugh asked for this feature, I thought it was going to be a breeze. 
Unfortunately, the Devil is in the details.

Reply:
Yes Satan is often hiding in Ember components :joy: No problem, thanks for the explanation!